experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp

```diff
@@ -0,0 +1,661 @@
+  if (failed(TimingDatabase::readFromJSON(timingModels, timingDB)))
+    signalPassFailure();
+
+  mlir::ModuleOp mod = getOperation();
+  for (handshake::FuncOp funcOp : mod.getOps<handshake::FuncOp>()) {
+
+    std::unordered_map<unsigned, llvm::SetVector<unsigned>> cfdfcBBLists;
+    std::unordered_map<unsigned, float> IIs;
+
+    // Extract CFDFCs and II from the Attributes
+    auto throughputAttr =
+        getDialectAttr<handshake::CFDFCThroughputAttr>(funcOp);
+    auto cfdfcAttr = getDialectAttr<handshake::CFDFCToBBListAttr>(funcOp);
+
+    if (throughputAttr == nullptr || cfdfcAttr == nullptr)
+      continue;
+
+    DictionaryAttr throughputDict = throughputAttr.getThroughputMap();
+    DictionaryAttr cfdfcDict = cfdfcAttr.getCfdfcMap();
+
+    // Convert Attribute into better usable data structure
+    for (const mlir::NamedAttribute attr : cfdfcDict) {
+      ArrayAttr bbList = llvm::dyn_cast<ArrayAttr>(attr.getValue());
+      llvm::SetVector<unsigned> cfdfcBBs;
+      for (auto bb : bbList)
+        cfdfcBBs.insert(bb.cast<IntegerAttr>().getUInt());
+
+      unsigned index;
+      if (attr.getName().getValue().getAsInteger(10, index))
+        signalPassFailure();
+
+      cfdfcBBLists.insert({index, cfdfcBBs});
+    }
+
+    // Extract II from Attribute
+    for (const NamedAttribute attr : throughputDict) {
+      FloatAttr throughput = llvm::dyn_cast<FloatAttr>(attr.getValue());
+      IIs.insert({std::stoi(attr.getName().str()),
+                  round(1 / throughput.getValueAsDouble())});
+    }
+
+    // Size LSQs for each CFDFC
+    for (auto &entry : cfdfcBBLists) {
+      // If there is no II corresponding to the CFDFC skip it
+      // Something is wrong with the attributes in that case
+      if (IIs.find(entry.first) == IIs.end())
+        continue;
+
+      std::optional<LSQSizingResult> result = sizeLSQsForCFDFC(
+          funcOp, entry.second, timingDB, IIs.at(entry.first), collisions);
+      if (result)
+        sizingResults.push_back(result.value());
+    }
+
+    // Extract maximum Queue sizes for each LSQ
+    std::map<mlir::Operation *, std::tuple<unsigned, unsigned>>
+        maxLoadStoreSizes;
+    for (auto &result : sizingResults) {
+      for (auto &entry : result) {
+        int newMaxLoadSize =
+            std::max(std::get<0>(maxLoadStoreSizes[entry.first]),
+                     std::get<0>(entry.second));
+        int newMaxStoreSize =
+            std::max(std::get<1>(maxLoadStoreSizes[entry.first]),
+                     std::get<1>(entry.second));
+        maxLoadStoreSizes[entry.first] =
+            std::make_tuple(newMaxLoadSize, newMaxStoreSize);
+      }
+    }
+
+    // Set the maximum Queue sizes as attributes for backend
+    for (auto &maxLoadStoreSize : maxLoadStoreSizes) {
+      mlir::Operation *lsqOp = maxLoadStoreSize.first;
+      unsigned maxLoadSize = std::get<0>(maxLoadStoreSize.second);
+      unsigned maxStoreSize = std::get<1>(maxLoadStoreSize.second);
```

Comment:
Here (i think) you can do something like this instead:

```c++
for (auto [lsqOp, [maxLoadSize, maxStoreSize]] : maxLoadStoreSizes) {
...
}
```

Reply:
I don't think C++ supports nested unpacking, you may have to do

```cpp
for (auto &[lsqOp, maxSizes] : maxLoadStoreSizes) {
  unsigned maxLoadSize = std::get<0>(maxSizes.second);
  unsigned maxStoreSize = std::get<1>(maxSizes.second);
}
```